### PR TITLE
add fut/back compat tests for implied trait bounds

### PR DESCRIPTION
**Pull request by lcnr**
the `guard` test was tested to cause a segfault with `-Zchalk`, very nice

cc @nikomatsakis #44491 #25860

### CODE REVIEW

File: src/test/ui/implied-bounds/hrlt-implied-trait-bounds-guard.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// A test exploiting the bug behind #25860 except with
+// implied trait bounds which currently don't exist,
+//
+// please ping @lcnr if your changes end up causing `badboi` to compile.
```
Comment (Mark-Simulacrum) [inline src/test/ui/implied-bounds/hrlt-implied-trait-bounds-guard.rs:4]: nit: probably best to place any such comment nearer to the ERROR in badboi.

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me (mostly just rubber stamping) with nit fixed

**Comment by lcnr** [comment]
woops, thanks :+1:

@bors r=Mark-Simulacrum rollup

**Pull request merged**